src/core/connection_pool.c

```diff
@@ -0,0 +1,521 @@
+/*++
+
+    Copyright (c) Microsoft Corporation.
+    Licensed under the MIT License.
+
+Abstract:
+
+    Definitions for the MsQuic Connection Pool API, which allows clients to
+    create a pool of connections that are spread across RSS cores.
+
+--*/
+
+#include "precomp.h"
+
+#ifdef QUIC_CLOG
+#include "connection_pool.c.clog.h"
+#endif
+
+#define MAX_CONNECTION_POOL_RETRY_MULTIPLIER 2
+
+const char*
+QuicConnPoolAllocServerNameCopy(
+    _In_ const char* ServerName,
+    _In_ size_t ServerNameLength
+    )
+{
+    char* ServerNameCopy = CXPLAT_ALLOC_NONPAGED(ServerNameLength + 1, QUIC_POOL_SERVERNAME);
+        if (ServerNameCopy == NULL) {
+            QuicTraceEvent(
+                AllocFailure,
+                "Allocation of '%s' failed. (%llu bytes)",
+                "Server name",
+                ServerNameLength + 1);
+        } else {
+            CxPlatCopyMemory(ServerNameCopy, ServerName, ServerNameLength);
+            ServerNameCopy[ServerNameLength] = 0;
+        }
+
+    return ServerNameCopy;
+}
+
+_IRQL_requires_max_(PASSIVE_LEVEL)
+QUIC_STATUS
+QuicConnPoolGetStartingLocalAddress(
+    _In_ QUIC_ADDR* RemoteAddress,
+    _Out_ QUIC_ADDR* LocalAddress
+    )
+{
+    CXPLAT_SOCKET* Socket = NULL;
+    CXPLAT_UDP_CONFIG UdpConfig = { .RemoteAddress = RemoteAddress, };
+    QUIC_STATUS Status =
+        CxPlatSocketCreateUdp(MsQuicLib.Datapath, &UdpConfig, &Socket);
+    if (QUIC_SUCCEEDED(Status)) {
+        CXPLAT_DBG_ASSERT(Socket != NULL);
+        CxPlatSocketGetLocalAddress(Socket, LocalAddress);
+        CxPlatSocketDelete(Socket);
+    }
+
+    return Status;
+}
+
+_IRQL_requires_max_(PASSIVE_LEVEL)
+QUIC_STATUS
+QuicConnPoolGetInterfaceIndexForLocalAddress(
+    _In_ QUIC_ADDR* LocalAddress,
+    _Out_ uint32_t* InterfaceIndex
+    )
+{
+    CXPLAT_ADAPTER_ADDRESS* Addresses = NULL;
+    uint32_t AddressesCount;
+
+    *InterfaceIndex = 0;
```

Comment:
Unnecessary, so please remove. This variable is defined as an `_Out_` and `QUIC_STATUS` is annotated with `_Return_type_success_(return >= 0)`.

Reply:
We need to initialize it to 0 so that line 88 can detect that no interface was found.